### PR TITLE
Add #7889: Company rating via Server Admin Port

### PR DESCRIPTION
**Pull request by pon-droid**
## Motivation / Problem
Enhancing the admin port by sending company 'ranking' information as well. 

## Description

The ranking is according to the performance history of the company, the same as the company league table. The company rating is modified by calling _SendCompanyInfo_.

## Limitations
- Modifies company objects
- Duplicates sorting functionality with the company league table

I originally attempted to reuse the sorting done from the company league table GUI class, but then the admin port would have to wait for the player to open it up before getting rating information.

This is also my first PR for any project so there are definitely more things to be called out here, please do that. 

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/company_base.h [hunk 1]
```diff
@@ -79,6 +79,8 @@ struct CompanyProperties {
 
 	Year inaugurated_year;           ///< Year of starting the company.
 
+	byte rating;                     ///< Ranking nth of the company with 1st being best performing according to performance history.
```
Comment (TrueBrain) [inline src/company_base.h]: You create this variable, but only write and read to it in a single function. That means you don't actually need to allocate the variable here. Saves a bit of memory that is always allocated while it is only used in a very specific case :) But we will see in a bit that I think we don't even need to do this :)

File: src/network/network_admin.cpp [hunk 2]
```diff
@@ -333,11 +332,27 @@ NetworkRecvStatus ServerNetworkAdminSocketHandler::SendCompanyInfo(const Company
 	p->Send_uint8 (CeilDiv(c->months_of_bankruptcy, 3)); // send as quarters_of_bankruptcy
 
 	for (auto owner : c->share_owners) {
-		p->Send_uint8(owner);
+	  p->Send_uint8(owner);
```
Comment (TrueBrain) [inline src/network/network_admin.cpp]: OpenTTD uses tabs to to align statement. I think you replace them with spaces here.

File: src/network/network_admin.cpp [hunk 3]
```diff
@@ -333,11 +332,27 @@ NetworkRecvStatus ServerNetworkAdminSocketHandler::SendCompanyInfo(const Company
...
+	for(Company* company: Company::Iterate()){
+	  companies.emplace_back(company);
+	}
+
+	std::sort(companies.begin(), companies.end(), [](Company * const &c1, Company * const &c2){
+	  return c2->old_economy[0].performance_history < c1->old_economy[0].performance_history;
+	});
+
+	for(size_t t = 0; t < companies.size(); t++){
+	  companies[t]->rating = t+1;
```
Comment (TrueBrain) [inline src/network/network_admin.cpp]: I am wondering about this block of logic, if that is the right thing to do. We manually sort the performance history here, and deduce a rating out of that. But why not let the admin-port clients do that? And just send over the `performance_history` to them? Saves us a lot of complex calculations, gives them more information than they receive now, and avoids nasty problems like: if you now request two companies with a slight delay, they might have the exact same rating as they so happened to have changed `performance_history`.

If we would go for this `rating` solution, the logic here could use some cleverness. You now calculate the rating for all companies, while you are only interested in a single one. So why not just calculate the single one? For example, it could be something like this (this replaces line 338 till 349 in your code):

```
int rating = 1;
for (auto company: Company::Iterate()) {
  if (c->old_economy[0].performance_history < company->old_economy[0].performance_history) rating++;
}
```

Basically, we find all companies that have a higher rating than us, meaning we drop one rating. The result is the exact same, but without the two loops and the sorting :) At least .. I think it is the same :P My simple testing shows it is :D But you do need to test it first :)

File: src/network/network_admin.cpp [hunk 4]
```diff
@@ -333,11 +332,27 @@ NetworkRecvStatus ServerNetworkAdminSocketHandler::SendCompanyInfo(const Company
...
 	}
 
+	p->Send_uint8 (c->rating);
```
Comment (TrueBrain) [inline src/network/network_admin.cpp]: I have no clue what the impact is of just adding data to an admin-port packet. Do clients break? I think we have a version-number somewhere; it might be needed to bump that number. Maybe someone like @nielsmh knows?

**Review by TrueBrain — COMMENTED** [review]
Excellent start for a Pull Request to OpenTTD! Thank you for that!

File: src/network/network_admin.cpp (hunk 3 above)
Comment (nielsmh) [reply src/network/network_admin.cpp]: Yeah that algorithm is simpler, the current way of getting everything, sorting it, and then counting, is excessive.
It could also be written with a `std::count_if()` call.

File: src/network/network_admin.cpp (hunk 4 above)
Comment (nielsmh) [reply src/network/network_admin.cpp]: Each packet contains a 16 bit size field at the start, so clients that just ignore data at the end of the packet that they don't know about should keep working. I'm pretty sure clients are supposed to do that.
I think the protocol version or packet type id should only be changed if fields are removed/reordered/redefined.

**Comment by ldpl** [comment]
I don't think this is a good idea. OpenTTD already sends performance ratings over the admin port so clients can figure out ranking themselves if they want. And adding it to the game itself does nothing but bloat the codebase needlessly IMO. Neither does rating really belong to company info logically as it's an attribute of a company list, not a single company on its own.

Also, the company league table is pointless and needs some way to customize it with proper ratings, not spreading its stupidity into more places.

File: src/network/network_admin.cpp [hunk 5]
```diff
@@ -336,8 +335,15 @@ NetworkRecvStatus ServerNetworkAdminSocketHandler::SendCompanyInfo(const Company
 		p->Send_uint8(owner);
 	}
 
+
+	uint8 rating = std::count_if(Company::Iterate().begin(), Company::Iterate().end(), [&](auto company){
+	return c->old_economy[0].performance_history < company->old_economy[0].performance_history;});
+
+	p->Send_uint8 (rating);
```
Comment (TrueBrain) [inline src/network/network_admin.cpp:342]: I posted this in an earlier comment, I am still not sure about the answer here. But do we actually want to send the rating itself to the client? As they can fetch the performance themselves, and deduce the rating based on it.

One use-case I can imagine is that you only want to fetch the company-info every N minutes and display that on some website. In that case I think it is a valid case, as you don't want to bother with fetching the economy and doing the calculation yourself.

But that said, the one thing that annoys me slightly about this, is that if you fetch multiple companies over time, you can have two companies with the same rating just because the game-state updated between your requests. The `COMPANY_ECONOMY` packet solves this by always sending all companies at once (and also includes the performance rating).

An alternative could be to add the rating in that packet explicit. Albeit a bit silly, it wouldn't hurt either. But changing that packet in a backwards compatible manner is nearly impossible.

Owh, the choices ... :D

Either way, curious what other developers think. @nielsmh , given you replied in the original issue too, how do you view this?

File: src/network/network_admin.cpp [hunk 6]
```diff
@@ -137,7 +137,6 @@ NetworkRecvStatus ServerNetworkAdminSocketHandler::SendError(NetworkErrorCode er
 
 	p->Send_uint8(error);
 	this->SendPacket(p);
-
```
Comment (TrueBrain) [inline src/network/network_admin.cpp:140]: Please in PRs always try to minimize the change. Removing these whiteline (as for the 2 others in this PR) carries no weight, and only distracts from your change. So it is better to remove them.

Otherwise we have constant changes in newlines in PRs, depending on the mood of the author ;) :D

File: src/network/network_admin.cpp [hunk 7]
```diff
@@ -336,8 +335,15 @@ NetworkRecvStatus ServerNetworkAdminSocketHandler::SendCompanyInfo(const Company
 		p->Send_uint8(owner);
 	}
 
+
+	uint8 rating = std::count_if(Company::Iterate().begin(), Company::Iterate().end(), [&](auto company){
```
Comment (TrueBrain) [inline src/network/network_admin.cpp:339]: a bit lower is a function `SendCompanyUpdate`, which sends the same info as this packet, minus some immutable fields (no clue how we ever were like: yeah, that is a great idea! Let's have two nearly identical packets, but .. here we are :D). As rating can change, and if we go for this change, I think this code should also be added there. Just to avoid confusion with any admin-port user :)

File: src/network/network_admin.cpp [hunk 8]
```diff
@@ -336,8 +335,15 @@ NetworkRecvStatus ServerNetworkAdminSocketHandler::SendCompanyInfo(const Company
...
+	return c->old_economy[0].performance_history < company->old_economy[0].performance_history;});
```
Comment (TrueBrain) [inline src/network/network_admin.cpp:340]: Coding style is a bit off here. I suggest something like this:

```
uint8 rating = std::count_if(Company::Iterate().begin(), Company::Iterate().end(), [&](auto company) {
    return c->old_economy[0].performance_history < company->old_economy[0].performance_history;
});
```

That way it is a bit more clear where the body of the lambda starts and ends :)

File: src/network/network_admin.cpp (hunk 5 above)
Comment (nielsmh) [reply src/network/network_admin.cpp:342]: When I wrote the reply in the original issue, I didn't realize that it was about sending the ranking position, and that the information for a client to deduce that was already available to the client.

Seeing this now, I'm very inclined to say close this and the original issue.

**Comment by TrueBrain** [comment]
I am really sorry @pon-droid , but it seems we let you astray. As per https://github.com/OpenTTD/OpenTTD/pull/9973#discussion_r957633496, the original issue should have been closed with: already implemented, and I will do as such in a moment.

But don't despair! There is nothing wrong with your PR itself. And tons of other small issues that require attention. So if you are up for it, maybe pick up one of the other small issues? :D

Tnx, and sorry :)
